// _@tarojs_webpack-runner@1.2.12@@tarojs/webpack-runner/dist/config/base.conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Chain = require("webpack-chain");
const util_1 = require("../util");
const chain_1 = require("../util/chain");
exports.default = () => {
    const chain = new Chain();
    chain.merge({
        module: {
            rule: {
                jsx: {
                    test: /\.jsx?$/,
                    use: {
                        babelLoader: chain_1.getBabelLoader([chain_1.defaultBabelLoaderOption])
                    }
                },
                media: {
                    test: /\.(mp4|webm|ogg|mp3|wav|flac|aac)(\?.*)?$/,
                    use: {
                        urlLoader: chain_1.getUrlLoader([chain_1.defaultMediaUrlLoaderOption])
                    }
                },
                font: {
                    test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
                    use: {
                        urlLoader: chain_1.getUrlLoader([chain_1.defaultFontUrlLoaderOption])
                    }
                },
                image: {
                    test: /\.(png|jpe?g|gif|bpm|svg)(\?.*)?$/,
                    use: {
                        urlLoader: chain_1.getUrlLoader([chain_1.defaultImageUrlLoaderOption])
                    }
                }
            }
        },
        resolve: {
            extensions: ['.js', '.jsx', '.ts', '.tsx'],
            mainFields: ['main:h5', 'main', 'module'],
            symlinks: true,
            modules: [
                path.join(util_1.appPath, 'node_modules'),
                'node_modules'
            ]
        },
        resolveLoader: {
            modules: [path.join(util_1.getRootPath(), 'node_modules')]
        }
    });
    return chain;
};
